README.md
# illume 

## A Widget-based Game Engine for Flutter 

Illume is a simple Flutter game engine which uses widgets to create game objects instead of sprites 
in normal game engines. This allows the app and game part to be quite integrated and use a lot of 
common components without needing graphics work to create sprites and backgrounds (which was honestly 
my main motivation, since I don't know anything about graphic design). It's also quite easy to create 
and manage the game state as shown in the features ahead. A lot of functionality still needs to be 
implemented including physics - only normal box-based collision is currently implemented.

Some parts were inspired by other game engines, primarily [Flame](https://github.com/flame-engine/flame).

DISCLAIMER: Since widgets are heavier than sprites drawn on a canvas, this engine is going to be 
less performant than others. It is in a very early stage and I plan to add a lot of features but 
this is primarily due to my interest in having widget-based game engines -- any serious game 
development should still be carried out with more mature engines like [Flame](https://github.com/flame-engine/flame).
This is just meant as an experiment to satisfy my curiosity. I quite like - and still heavily use the 
main Flutter game engines whose development teams I respect a lot. 

## Features

### Use Widgets directly to build your game objects

```dart
class FlappyWidget extends GameObject {

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Text('Demo'),
    );
  }

  @override
  void init() {}

  @override
  void onCollision(List<Collision> collisions) { }

  @override
  void onScreenSizeChange(Vector2 size) {}

  @override
  void update(Duration delta) {}

  //...
}
```

### Easy game development without large separation between widgets and games 

You can add a game to your app simply by adding an `Illume` widget and an `IllumeController`. (More 
info given in usage section)

```dart
class _MyHomePageState extends State<MyHomePage> {
  IllumeController gameController = IllumeController();

  FlappyWidget flappyWidget = FlappyWidget();

  @override
  void initState() {
    super.initState();
    gameController.startGame();
    gameController.gameObjects.add(flappyWidget);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Illume(
        illumeController: gameController,
      ),
    );
  }
}

// Definition of game object given later
```

### `IllumeController` allows easy manipulation of game state

```dart
  IllumeController gameController = IllumeController();

  gameController.startGame();
  gameController.gameObjects.add(DemoObject());
  gameController.stopGame();
  // etc
```

The controller is also available for use in all game objects so no need for an abundance of callbacks.

### Collision detection

Collision detection is enabled by default for all objects which can be turned off. Every object is 
notified of collisions as well as given the `Rect` of intersection so collisions can be better 
understood. This is a very early prototype so no complex physics (Box2d-ish) exist but I hope to 
add this in the future.

```dart
class FlappyWidget extends GameObject {

  //...
  
  @override
  void onCollision(List<Collision> collisions) { 
    // This gives a list of objects which are colliding with this one 
    // as well as the rect of intersection.
  }

  //...
}
```

### Gesture Detection

This is technically NOT a feature of illume since we can use the default Flutter gesture detecting 
widgets like the `GestureDetector` to power our apps. Check the example provided in the repo for more 
but here is a short demo: 

```dart
@override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        flappyWidget.jump();
      },
      child: Scaffold(
        appBar: AppBar(
          title: Text('Illume Demo'),
        ),
        body: Illume(
          illumeController: gameController,
        ),
      ),
    );
  }
```

## Usage

### Step 1

Add the dependency first: 

```yaml
dependencies:
  illume: ^0.1.0
```

### Step 2

Add the `Illume` widget and associated `IllumeController` wherever you need to create a game:

```dart
IllumeController gameController = IllumeController();

@override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Illume Demo'),
      ),
      body: Illume(
        illumeController: gameController,
      ),
    );
  }
```

### Step 3

Define your `GameObject` by extending the class:

```dart
class FlappyWidget extends GameObject {
  var velocity = 0.0;
  var acceleration = 0.2;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Text('Demo'),
    );
  }

  @override
  void init() {
    size = Vector2.all(50);
    alignment = GameObjectAlignment.center;
    position = (gameSize / 2);
  }

  @override
  void onCollision(List<Collision> collisions) {
    illumeController.stopGame();
  }

  @override
  void onScreenSizeChange(Vector2 size) {
    // This is a quick demo but you really should shift your positions in a
    // real world app or at least lock orientation.
  }

  @override
  void update(Duration delta) {
    position += Vector2(0, velocity);
    velocity = velocity + acceleration;
  }

  void jump() {
    velocity = -5;
  }
}
```

### Note: In the object there are various properties - 

* `gameSize` gives the max size for the game. 
* `alignment` allows you to align the center to the center of the widget or the top left corner of 
the widget.
* `position` is the position of the widget.
* `size` is the size of the widget.
* `illumeController` is the game controller for the game. 

### Step 4

Add the object to the game and start the game:

```dart
  FlappyWidget flappyWidget = FlappyWidget();

  @override
  void initState() {
    super.initState();
    gameController.startGame();
    gameController.gameObjects.add(flappyWidget);
  }
```

Check out the full example for the demo flappy bird game (This is a very basic demo for now, I'll 
add a far better one later).

CHANGELOG.md
## 0.2.1

* Added initialization check when rebuilding

## 0.2.0 

* Added option to rebuild Widget

## 0.1.2 [17/04/2022]

* Better docs 
* README corrections

## 0.1.1 [08/12/2021]

README corrections

## 0.1.0 [08/12/2021]

Initial release

Contains game objects and collisions


test/illume_test.dart
// import 'package:flutter_test/flutter_test.dart';
//
// import 'package:illume/illume.dart';

void main() {}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

// import 'package:flutter/material.dart';
// import 'package:flutter_test/flutter_test.dart';

void main() {
  // testWidgets('Counter increments smoke test', (WidgetTester tester) async {
  //   // Build our app and trigger a frame.
  //   await tester.pumpWidget(const MyApp());
  //
  //   // Verify that our counter starts at 0.
  //   expect(find.text('0'), findsOneWidget);
  //   expect(find.text('1'), findsNothing);
  //
  //   // Tap the '+' icon and trigger a frame.
  //   await tester.tap(find.byIcon(Icons.add));
  //   await tester.pump();
  //
  //   // Verify that our counter has incremented.
  //   expect(find.text('0'), findsNothing);
  //   expect(find.text('1'), findsOneWidget);
  // });
}


lib/src/illume.dart
import 'package:flutter/material.dart';
import 'package:illume/illume.dart';
import 'package:illume/src/extensions.dart';

/// The Widget responsible for creating an Illume game
/// The [illumeController] is responsible for holding the game objects
class Illume extends StatefulWidget {
  final IllumeController illumeController;

  const Illume({
    Key? key,
    required this.illumeController,
  }) : super(key: key);

  @override
  _IllumeState createState() => _IllumeState();
}

class _IllumeState extends State<Illume> {
  Offset lastSize = const Offset(0, 0);
  Map<GameObject, Widget> widgetCache = {};

  @override
  Widget build(BuildContext context) {
    return ValueListenableBuilder<List<GameObject>>(
      valueListenable: widget.illumeController,
      builder: (context, value, child) {
        var objects = getVisibleGameObjects();

        return LayoutBuilder(builder: (context, constraints) {
          var newSize = Offset(constraints.maxWidth, constraints.maxHeight);
          if (newSize != lastSize) {
            _changeSize(newSize);
            lastSize = newSize;
          }

          return Stack(
            children: objects.map(
              (e) {
                var position = e.getPositionRect();
                return Positioned(
                  top: position.topLeft.dy,
                  left: position.topLeft.dx,
                  child: SizedBox(
                    width: e.size[0],
                    height: e.size[1],
                    child: getObjectWidget(e),
                  ),
                );
              },
            ).toList(),
          );
        });
      },
    );
  }

  Widget getObjectWidget(GameObject e) {
    if (widgetCache[e] != null && !e.rebuildWidgetIfNeeded) {
      return widgetCache[e]!;
    }

    var newWidget = e.build(context);
    widgetCache[e] = newWidget;

    if (widgetCache.length != widget.illumeController.gameObjects.length) {
      var removeList = widgetCache.keys
          .where((e) => !widget.illumeController.gameObjects.contains(e))
          .toList();
      removeList.forEach(widgetCache.remove);
    }

    if (e.rebuildWidgetIfNeeded) {
      e.rebuildWidgetIfNeeded = false;
    }

    return newWidget;
  }

  List<GameObject> getVisibleGameObjects() {
    var objects = widget.illumeController.gameObjects;
    objects.removeWhere((e) => !e.visible);
    return objects;
  }

  void _changeSize(Offset size) {
    for (var e in widget.illumeController.gameObjects) {
      e.gameSize = Vector2(size.dx, size.dy);
      if (!e.initialised) {
        e.illumeController = widget.illumeController;
        e.init();
        e.initialised = true;
      }
      e.onScreenSizeChange(size.toVector2);
    }
  }
}


lib/src/illume_controller.dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:illume/src/components/game_object.dart';
import 'package:illume/src/extensions.dart';

import 'components/collision.dart';

/// The controller for the game which contains the game objects and allows
/// controlling the start/stop/pause mechanism for the game
class IllumeController extends ValueNotifier<List<GameObject>> {
  /// Game objects present in the game
  late List<GameObject> gameObjects;

  /// Create default [IllumeController]
  factory IllumeController() => IllumeController._([]);

  /// Create [IllumeController] from game objects
  factory IllumeController.fromObjects(List<GameObject> objects) =>
      IllumeController._(objects);

  IllumeController._(this.gameObjects) : super(gameObjects) {
    _ticker = Ticker(_update);
  }

  DateTime? gameLoopStart;
  DateTime? lastTick;
  late Ticker _ticker;

  ValueChanged<GameObject>? onObjectRemoved;

  void _update(Duration elapsed) {
    lastTick = DateTime.now();
    for (var e in gameObjects) {
      e.update(elapsed);
    }
    _checkCollisions();

    notifyListeners();
  }

  void startGame() {
    gameLoopStart = DateTime.now();
    lastTick = DateTime.now();
    _ticker.start();
  }

  void pause() {
    _ticker.muted = true;
  }

  void resume() {
    _ticker.muted = false;
    if (!_ticker.isActive) {
      startGame();
    }
  }

  void stopGame() {
    _ticker.stop();
  }

  void rebuildWidgetsWhereNeeded() {
    notifyListeners();
  }

  @override
  void dispose() {
    super.dispose();
    _ticker.dispose();
  }

  get gameInProgress => _ticker.isTicking && !_ticker.muted;

  void _checkCollisions() {
    if (gameObjects.length < 2) return;

    var objects = gameObjects.toList();
    objects.removeWhere((e) => !e.collidable);

    var intersectionMap = <GameObject, List<Collision>>{};
    for (var e in objects) {
      intersectionMap[e] = [];
    }

    for (int i = 0; i < objects.length - 1; i++) {
      for (int j = i + 1; j < objects.length; j++) {
        var obj1 = objects[i];
        var obj2 = objects[j];
        var rect1 = Rect.fromPoints((obj1.position - (obj1.size / 2)).toOffset,
            (obj1.position + (obj1.size / 2)).toOffset);
        var rect2 = Rect.fromPoints((obj2.position - (obj2.size / 2)).toOffset,
            (obj2.position + (obj2.size / 2)).toOffset);
        if (rect1.overlaps(rect2)) {
          intersectionMap[obj1]!.add(Collision(rect1.intersect(rect2), obj2));
          intersectionMap[obj2]!.add(Collision(rect2.intersect(rect1), obj1));
        }
      }
    }

    intersectionMap.forEach((key, value) {
      if (value.isNotEmpty) {
        key.onCollision(value);
      }
    });
  }
}


lib/src/components/game_object.dart
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math.dart';
import 'package:illume/src/extensions.dart';

import '../illume_controller.dart';
import 'collision.dart';

/// Defines the alignment of the [GameObject]
/// If set to [GameObjectAlignment.topLeft], the position of the object is calculated wrt to the top left corner
/// If set to [GameObjectAlignment.center], the position of the object is calculated wrt to the center
enum GameObjectAlignment {
  center,
  topLeft,
}

/// Class defining a game object for Illume
abstract class GameObject {
  Vector2 position = Vector2.zero();
  Vector2 size = Vector2.zero();
  Vector2 gameSize = Vector2.zero();
  bool visible = true;
  bool collidable = true;
  GameObjectAlignment alignment = GameObjectAlignment.center;
  bool initialised = false;
  bool rebuildWidgetIfNeeded = false;
  late IllumeController illumeController;

  void init();

  Widget build(BuildContext context);

  void update(Duration delta);

  void onScreenSizeChange(Vector2 size);

  void onCollision(List<Collision> collisions);

  Rect getPositionRect() {
    switch (alignment) {
      case GameObjectAlignment.center:
        return Rect.fromPoints(position.toOffset - (size.toOffset / 2.0),
            position.toOffset + (size.toOffset / 2.0));
      case GameObjectAlignment.topLeft:
        return Rect.fromPoints(
            position.toOffset, position.toOffset + size.toOffset);
    }
  }

  void rebuildWidget() {
    rebuildWidgetIfNeeded = true;
    if (initialised) {
      illumeController.rebuildWidgetsWhereNeeded();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:illume/illume.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

/// Quick demo for vertical screens
class _MyHomePageState extends State<MyHomePage> {
  IllumeController gameController = IllumeController();

  FlappyWidget flappyWidget = FlappyWidget();
  Wall wall = Wall(200, false);
  Wall wall2 = Wall(400, true);

  @override
  void initState() {
    super.initState();
    gameController.startGame();
    gameController.gameObjects.addAll([flappyWidget, wall, wall2]);
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        flappyWidget.jump();
      },
      child: Scaffold(
        appBar: AppBar(
          title: const Text('Illume Demo'),
        ),
        body: Stack(
          children: [
            gradient,
            Illume(
              illumeController: gameController,
            ),
          ],
        ),
      ),
    );
  }

  var gradient = Container(
    decoration: const BoxDecoration(
      gradient: LinearGradient(
          colors: [Colors.teal, Colors.deepPurple],
          stops: [0.0, 1.0],
          begin: Alignment.topLeft,
          end: Alignment.bottomRight),
    ),
  );
}

class FlappyWidget extends GameObject {
  var velocity = 0.0;
  var acceleration = 0.2;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: const Text('Demo'),
    );
  }

  @override
  void init() {
    size = Vector2.all(50);
    alignment = GameObjectAlignment.center;
    position = (gameSize / 2);
  }

  @override
  void onCollision(List<Collision> collisions) {
    illumeController.stopGame();
  }

  @override
  void onScreenSizeChange(Vector2 size) {
    // This is a quick demo but you really should shift your positions in a
    // real world app or at least lock orientation.
  }

  @override
  void update(Duration delta) {
    position += Vector2(0, velocity);
    velocity = velocity + acceleration;
  }

  void jump() {
    velocity = -5;
  }
}

class Wall extends GameObject {
  int initialDistance;
  bool vert;

  Wall(
    this.initialDistance,
    this.vert,
  );

  var velocity = -2.0;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.green,
      child: const Text('Demo'),
    );
  }

  @override
  void init() {
    size = Vector2(50, gameSize[1] / 2);
    alignment = GameObjectAlignment.center;
    position = Vector2(
      (gameSize[0] / 2) + initialDistance,
      (vert ? 1 : 3) * gameSize[1] / 4,
    );
  }

  @override
  void onCollision(List<Collision> collisions) {
    illumeController.stopGame();
  }

  @override
  void onScreenSizeChange(Vector2 size) {
    // This is a quick demo but you really should shift your positions in a
    // real world app or at least lock orientation.
  }

  @override
  void update(Duration delta) {
    position += Vector2(velocity, 0);
    if (position[0] < -50) {
      position[0] = gameSize[0] + 25;
    }
  }
}
